### PR TITLE
Cuda lfp implementation

### PR DESCRIPTION
**Pull request by thorulf4**

### CODE REVIEW

File: src/bit_cuda/analysis.cu [hunk 1]
```diff
@@ -89,8 +105,16 @@ void bit_cuda::execute_analysis(Node* nodes, int node_count) {
...
     }
 
+    if(extra_transfer_count > 0){
+        cudaStatus = cudaMemcpy(dev_extra_transfers, transfers, sizeof(Transfer)*extra_transfer_count, cudaMemcpyHostToDevice);
+        if (cudaStatus != cudaSuccess) {
+            fprintf(stderr, "Copy false to gpu failed");
+            goto Error;
+        }
+    }
+
```
Comment (NicklasHugoy) [inline src/bit_cuda/analysis.cu:115]: Why are we doing this twice?

File: src/bit_cuda/analysis.cu [hunk 2]
```diff
@@ -144,8 +143,10 @@ Error:
     if(dev_extra_transfers != nullptr){
         cudaFree(dev_extra_transfers);
     }
+
+std::cout << "Test1\n";
```
Comment (NicklasHugoy) [inline src/bit_cuda/analysis.cu:147]: Remove?

File: src/bit_cuda/bit_cuda_transformer.h [hunk 3]
```diff
@@ -2,8 +2,8 @@
 #include "analysis.h"
 #include <var_visitor.h>
 
-#define RETURN_VAR "£return"
-#define TAINT_VAR "£"
+#define RETURN_VAR "$return"
+#define TAINT_VAR "$"
```
Comment (NicklasHugoy) [inline src/bit_cuda/bit_cuda_transformer.h:6]: Already included from cfg.h

File: src/bit_cuda/bit_cuda_transformer.h [hunk 4]
```diff
@@ -16,13 +16,17 @@ class BitCudaTransformer : public CfgVisitor<LatticeType>
 
     BitCudaTransformer(std::set<std::string> progVariables){
         int i = 0;
-        variables[RETURN_VAR] = i++;
         variables[TAINT_VAR] = i++;
+        variables[RETURN_VAR] = i++;
+
+        std::cout << TAINT_VAR << " -> " << variables[TAINT_VAR] << " : " << (1 << variables[TAINT_VAR])  <<'\n';
+        std::cout << RETURN_VAR << " -> " << variables[RETURN_VAR] << " : " << (1 << variables[RETURN_VAR])  <<'\n';
```
Comment (NicklasHugoy) [inline src/bit_cuda/bit_cuda_transformer.h:23]: Delete test prints

File: src/bit_cuda/bit_cuda_transformer.h [hunk 5]
```diff
@@ -16,13 +16,17 @@ class BitCudaTransformer : public CfgVisitor<LatticeType>
...
 
         std::set<std::string>::iterator it;
         for (it=progVariables.begin(); it!=progVariables.end(); it++){
             if(*it != TAINT_VAR && *it != RETURN_VAR){
                 variables[*it] = i++;
+                std::cout << *it << " -> " << variables[*it] << " : " << (1 << variables[*it])  <<'\n';
```
Comment (NicklasHugoy) [inline src/bit_cuda/bit_cuda_transformer.h:29]: Delete test prints

**Review by NicklasHugoy — CHANGES_REQUESTED** [review]

File: src/bit_cuda/bit_cuda_transformer.h [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        bit_cuda::Node& node_struct = nodes.emplace_back();
+    }
+
+    void visit_functionEntry(FunctionEntryNode<LatticeType>& node) { 
+        node_to_index[&node] = nodes.size();
+        bit_cuda::Node& node_struct = nodes.emplace_back();
+
+        if(node.arguments.size() >= 1){
+            bit_cuda::Transfer& last = node_struct.transfer;
+            fill_with_variable_indices(node_struct.transfer.rhs, node.arguments[0]);
```
Comment (NicklasHugoy) [inline src/bit_cuda/bit_cuda_transformer.h:64]: `node_struct.transfer.x` is never assigned to anything

File: src/main.cpp [hunk 7]
```diff
@@ -50,10 +53,31 @@ void cpu_multi_taint_analysis(ScTransformer<SourcedTaintState> program){
...
+void bit_cuda_analysis(ScTransformer<std::set<std::string>> program){
+    time_func("Variable reduction: ", 
+                reduce_variables<std::set<std::string>>, program.entryNodes);
+    auto transformer = time_func<BitCudaTransformer<std::set<std::string>>>("Gpu structure transformation: ", 
+                transform_bit_cuda<std::set<std::string>>, program.nodes);
+    
+
+    if(transformer.extra_transfers.size() >0 ){
+        time_func("Least fixed point algorithm: ",
+                bit_cuda::execute_analysis, &transformer.nodes[0], transformer.nodes.size(), &*transformer.extra_transfers.begin(), transformer.extra_transfers.size());
```
Comment (NicklasHugoy) [inline src/main.cpp:65]: Use `&transformer.extra_transfers[0]` instead?

**Review by ZcortezZ — APPROVED** [review]

**Pull request merged**